src/components/Modal/Modal.tsx

```diff
@@ -59,13 +55,40 @@ export const Modal = ({
     return () => {
       document.removeEventListener("keydown", keyDown);
     };
-  }, [close]);
+  });
+
+  const modalRef = React.useRef();
+  const handleTabKey = (e) => {
+    const focusableModalElements = modalRef.current.querySelectorAll(
+      'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
+    );
+    const firstElement = focusableModalElements[0];
+    const lastElement =
+      focusableModalElements[focusableModalElements.length - 1];
+
+    if (!e.shiftKey && document.activeElement !== firstElement) {
+      firstElement.focus();
+      return e.preventDefault();
+    }
+
+    if (e.shiftKey && document.activeElement !== lastElement) {
+      lastElement.focus();
+      e.preventDefault();
+    }
+  };
+
+  const keyListenersMap = new Map([
+    [27, close],
+    [9, handleTabKey],
+  ]);
 
   return (
     <div
       className={classNames("p-modal", className)}
       onClick={close}
       {...wrapperProps}
+      role="dialog"
```

Comment:
This will have duplicate nested elements with the role `dialog` (see the `section` element below).